jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/ThreadIdPool.java

```diff
@@ -0,0 +1,234 @@
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.util.thread;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReferenceArray;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import org.eclipse.jetty.util.ProcessorUtils;
+import org.eclipse.jetty.util.component.Dumpable;
+import org.eclipse.jetty.util.component.DumpableCollection;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A fixed sized pool of items that uses ThreadId to avoid contention.
+ * This class can be used, instead of a {@link ThreadLocal}, when pooling items
+ * that are expensive to create, but only used briefly in the scope of a single thread.
+ * It is safe to use with {@link org.eclipse.jetty.util.VirtualThreads}, as unlike a {@link ThreadLocal} pool,
+ * the number of items is limited.
+ * <p>This is a light-weight version of {@link org.eclipse.jetty.util.ConcurrentPool} that is best used
+ * when items do not reserve an index in the pool even when acquired.
+ * @see org.eclipse.jetty.util.ConcurrentPool
+ */
+public class ThreadIdPool<E> implements Dumpable
+{
+    private static final Logger LOG = LoggerFactory.getLogger(ThreadIdPool.class);
+
+    private final AtomicReferenceArray<E> _items;
+
+    public ThreadIdPool()
+    {
+        this(-1);
+    }
+
+    public ThreadIdPool(int capacity)
+    {
+        _items = new AtomicReferenceArray<>(calcCapacity(capacity));
+        if (LOG.isDebugEnabled())
+            LOG.debug("{}", this);
+    }
+
+    private static int calcCapacity(int capacity)
+    {
+        if (capacity >= 0)
+            return capacity;
+        return 2 * ProcessorUtils.availableProcessors();
+    }
+
+    /**
+     * @return the maximum number of items
+     */
+    public int capacity()
+    {
+        return _items.length();
+    }
+
+    /**
+     * @return the number of items available
+     */
+    public int size()
+    {
+        int available = 0;
+        for (int i = _items.length(); i-- > 0; )
+            if (_items.getPlain(i) != null)
+                available++;
+        return available;
+    }
+
+    /**
+     * Offer an item to the pool.
+     * @param e The item to offer
+     * @return The index the item was added at or -1, if it was not added
+     * @see #remove(Object, int) 
+     */
+    public int offer(E e)
+    {
+        int capacity = capacity();
+        if (capacity > 0)
+        {
+            int index = (int)(Thread.currentThread().getId() % capacity);
+            for (int i = capacity; i-- > 0; )
```

Comment:
I prefer a forward loop, as it is only used to bound the attempts -- it's easier to read.
Also in the other loops in other methods.

Reply:
@sbordet backwards loops are faster as the decrement and 0 check is one less instruction that increment and compare. Also the -- saves an extra operation for the index counting from 1.   This is hot path stuff, so every instruction counts.  We have previously benchmarked these loops and they were faster.
The readability is not that bad... if you can't work out such a loop...

Reply:
Or the JIT does it 😃

Reply:
@gregw time to update: https://arnaudroger.github.io/blog/2017/06/15/forward-vs-backward-loop.html 😈

Reply:
It didn't last time we tested.  might do so now, but I still like the form of it.  The for loop is simpler, with only 2 clauses rather than 3.   But if you want to write a JMH to prove that it is not necessary, then go for it!

Reply:
@gregw others have done the JMH benchmark and looked at the generated JIT machine code, read the article.
Forward is faster and reads better.